vault/logical_system.go

```diff
@@ -725,6 +729,140 @@ func (b *SystemBackend) handlePluginReloadUpdate(ctx context.Context, req *logic
+		cpu := float32(d.Get("cpu").(float64))
+		if cpu == 0 {
+			cpu = pluginruntimeutil.DefaultCPU
+		}
+		memory := uint64(d.Get("memory").(int64))
+		if memory == 0 {
+			memory = pluginruntimeutil.DefaultMemory
+		}
+		err := b.Core.pluginRuntimeCatalog.Set(ctx, runtimeName, runtimeType, ociRuntime, cgroupParent, cpu, memory)
+		if err != nil {
+			return logical.ErrorResponse(err.Error()), nil
+		}
+	default:
+		logical.ErrorResponse(fmt.Sprintf("%s is not a supported plugin runtime type", runtimeTypeStr))
+	}
+	return nil, nil
+}
+
+func (b *SystemBackend) handlePluginRuntimeCatalogDelete(ctx context.Context, _ *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	runtimeName := d.Get("name").(string)
+	if runtimeName == "" {
+		return logical.ErrorResponse("missing plugin runtime name"), nil
+	}
+
+	runtimeTypeStr := d.Get("type").(string)
+	if runtimeTypeStr == "" {
+		return logical.ErrorResponse("missing plugin runtime type"), nil
+	}
+
+	runtimeType, err := consts.ParsePluginRuntimeType(runtimeTypeStr)
+	if err != nil {
+		return logical.ErrorResponse(err.Error()), nil
+	}
+
+	err = b.Core.pluginRuntimeCatalog.Delete(ctx, runtimeName, runtimeType)
+	if err != nil {
+		return logical.ErrorResponse(err.Error()), nil
+	}
+	return nil, nil
+}
+
+func (b *SystemBackend) handlePluginRuntimeCatalogRead(ctx context.Context, _ *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	runtimeName := d.Get("name").(string)
+	if runtimeName == "" {
+		return logical.ErrorResponse("missing plugin runtime name"), nil
+	}
+
+	runtimeTypeStr := d.Get("type").(string)
+	if runtimeTypeStr == "" {
+		return logical.ErrorResponse("missing plugin runtime type"), nil
+	}
+
+	runtimeType, err := consts.ParsePluginRuntimeType(runtimeTypeStr)
+	if err != nil {
+		return logical.ErrorResponse(err.Error()), nil
+	}
+
+	conf, err := b.Core.pluginRuntimeCatalog.Get(ctx, runtimeName, runtimeType)
+	if err != nil {
+		return logical.ErrorResponse(err.Error()), nil
+	}
+	return &logical.Response{Data: map[string]interface{}{
+		"name":          conf.Name,
+		"type":          conf.Type.String(),
+		"oci_runtime":   conf.OCIRuntime,
+		"cgroup_parent": conf.CgroupParent,
+		"cpu":           conf.CPU,
+		"memory":        conf.Memory,
+	}}, nil
+}
+
+func (b *SystemBackend) handlePluginRuntimeCatalogTypedList(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	runtimeType, err := consts.ParsePluginRuntimeType(d.Get("type").(string))
+	if err != nil {
+		return nil, err
+	}
+
+	runtimes, err := b.Core.pluginRuntimeCatalog.List(ctx, runtimeType)
+	if err != nil {
+		return nil, err
+	}
+	sort.Strings(runtimes)
+	return logical.ListResponse(runtimes), nil
+}
+
+func (b *SystemBackend) handlePluginRuntimeCatalogUntypedList(ctx context.Context, _ *logical.Request, _ *framework.FieldData) (*logical.Response, error) {
+	data := make(map[string]interface{})
```

Comment:
Can we make this return just a list and not a map? It leaves us more flexibility for future changes to the API if we separate the data from the JSON format as much as possible, and each entry should already have the type recorded within it already anyway right?

Reply:
Would returning data with the following structure work?
```[
    {
      "name": "gvisor",
      "type": "container",
      "oci_runtime": "runsc",
      "cgroup_parent": "/cpulimit/",
      "cpu": 1,
      "memory": 10000
    },
    {
      "name": "other",
      "type": "otherType",
      "oci_runtime": "otherOCIRuntime",
      "cgroup_parent": "/cpulimit/",
      "cpu": 1,
      "memory": 10000
    }
  ]
```

Reply:
Yeah I think that looks perfect! 👍